{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## Analysis of Gene Expression Data via Arrays using Bioconductor/R - III"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Today, this notebook constitutes your in-class activity and homework. \n",
    "\n",
    "#### We will pick up from where we left off with the analysis of our gene expression data set from last time.\n",
    "\n",
    "#### We left off having performed our differential gene-expression analysis. However, our results were in terms of \"probes\" rather than genes; we'd like to be able to know what genes are actually of interest in our analysis. Today, we'll connect our analysis to annotation data so that we can obtain this information."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q1. Let's recover our pipline from last time.\n",
    "\n",
    "- Copy code from the in-class assignment Gene Expr. II, from: Q1. These are the steps that correspond to: (i) loading R libraries, (ii) loading the file that maps phenotype to cel files (i.e., the .csv file), (iii) storing the list of cel files available, (iv) reading the cel files.\n",
    "\n",
    "- Copy code from the in-class assignment Gene Expr. II, from: Q2 and Q3. These steps correspond to (i) the normalization step, and (ii) creating a box-plot of the normalized array data.\n",
    "\n",
    "- Copy code from the in-class assignment Gene Expr. II, from: Q4 and Q5. These steps correspond to (i) the group/design analysis matrices, (ii) the analysis, (iii) the empirical Bayes correction, (iv) the top table construction, and (v) print the top five results to screen. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
   ],
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q2. Load the annotation library package biomaRt in R. Provide and execute your R code below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
   ],
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q3. OK, now let's obtain some annotation information\n",
    "\n",
    "- create a variable called \"mymart\", which loads the appropriate library from the ensembl database for mouse.\n",
    "- obtain a list of the affy probe ids for the top 100 results variable \"tt\" above, stored in a variable called pidsTophits\n",
    "- print the first 5 probe entries from pidsTophits\n",
    "\n",
    "#### Provide and execute your R code below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
   ],
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q4. Continue obtaining annotation information. Note that getting all of the annotations will take some time!\n",
    "\n",
    "- obtain the following annotations from ensembl stored in a new variable called \"myannot\" for your list of affy probe ids:\n",
    "        probeids (for the array technology used for this experiment, affy_mouse430_2)\n",
    "        chromosome\n",
    "        start position\n",
    "        end position\n",
    "        mgi gene symbol (mouse)\n",
    "        entrez gene id\n",
    "        Description\n",
    "        \n",
    "- print out to screen for the first 5 entries in your table: probeid, chr, start, end, mgi, and description\n",
    "\n",
    "#### Provide and execute your R code below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
   ],
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q5. Output your annotations to file that reports everything but the description, without quotes, and separated by commas (useful for data parsing) to a file called tt-annot-forparse.csv\n",
    "\n",
    "#### Provide and execute your R code below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
   ],
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q5. Now that we have annotations, we need to merge them together with the table that contained our association information. However, notice that the annot matrix lists the probe ids in a different order than the statistical association table.\n",
    "\n",
    "- Create a table called \"tt_ids\" which includes data from the \"tt\" variable with the variable \"pidsTophits\", using the R function cbind(). Store into a new variable called tt_ids.\n",
    "- Rename the column containing pidTophits to affyprobeid in the variable tt_ids (hint: colnames).\n",
    "- print the first five entries of tt_ids to the screen.\n",
    "\n",
    "#### Provide and execute your R code below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
   ],
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q6. Now, we need to prepare and merge the annot file with tt_ids:\n",
    "\n",
    "- Rename the first column in the variable \"annot\" - the one that contains your probeid to \"affyprobeid\" using the R function colnames().\n",
    "- output the first 5 entries in the variable \"annot\" to ensure that the name has been changed\n",
    "- Use the R function merge() to aggregate the variable \"tt_ids\" with \"annot\", using affyprobeid as the key. Save the results into a new table called \"tt_plus_annot\"\n",
    "\n",
    "#### Provide and execute your R code below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
   ],
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q7. Let's sort the table to get the most significant results to the top of the dable. \n",
    "\n",
    "- use the R function order() to sort the the table by the key \"adj.P.Val\". Store the sorted table to a new variable called \"tt_annot_s\"\n",
    "- output the results of the first 5 row of \"tt_annot_s\"\n",
    "\n",
    "#### Provide and execute your R code below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
   ],
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q8. What are the genes listed in the top 10 results? "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "The top 20 differentially expressed probes map to the genes: Mbl2, Mafb, Rgs1, Ubd, Pglyrp2, Il7r, Gpnmb, Lpl, Lilra6, and Gm15922."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q9. How many unique genes are given in the your top results variable tt_annot_s? \n",
    "\n",
    "- TIP: Think through (in human terms) the specific tasks/function calls you will need to do perform the task. \n",
    "- Create a variable called \"uniqgenelist\" that lists your mgi gene symbols. (Hint: Use the variable \"tt_annot_s\".)\n",
    "- Double check the list of genes that you have constructed. Do all of the entries have a gene listed? \n",
    "- Create a new variable called \"uniqgenelist_filt\" that excludes entries that are \"NA\" or are empty (i.e., \"\").\n",
    "\n",
    "- Hint #2: length(), unique(), and is.na() functions may be useful to you.\n",
    "- Hint #3: the \"!\" (she-bang) character computes ths inverse of the function. e.g. \"!is.na(list)\" means in human terms \"which elements in list are **NOT** NA?\"\n",
    "- Hint #4: you can use logical operations (i.e., ==, !=) to match 'empty' entries (i.e., \"\"), even within []. Try it!\n",
    "\n",
    "#### Provide and execute your R code below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
   ],
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q10. Now, let us use WebGestalt to perform some ontology analysis using the unique set of genes created above.\n",
    "\n",
    "- Save the list of unique genes (filtered to exclude \"NA\"s) to a file called \"mygenelist.txt\", no quotes, row names, or column names\n",
    "\n",
    "#### Provide and execute your R code below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
   ],
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "#### Q11. Copy-paste the list of genes into WebGestalt. Does that tool find all of the genes that you provided? Why or why not? How could you fix this?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q12. What background gene list will you select? (Hint what ) Why is this important? \n",
    "\n",
    "- Find the annotation for the array data that you loaded. Hint: What is found in \"affyRaw\", the variable that stores your array data?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q13. Perform enrichment analyses and perform. Set the statistical method to \"Hypergeometric\", the correction to \"BH\" (Benjaminiâ€“Hochberg), and the significance level to 0.001.  Do you find any associations with this level of support or better? Biologically, do any of these categories make plausible or intuitive sense for the gene expression analysis that you performed? "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q14. Here, we selected the top 100 genes for enrichement analysis. Was this choice warrented? Why or why not? Having performed this analysis, what might you do differently?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Q11. One final item in your pipeline to finish it off complete.\n",
    "\n",
    "- run the R function sessionInfo() to record the functions and versions used for your analysis.\n",
    "\n",
    "#### Provide and execute your R code below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
   ],
   "source": [
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Congrats! You have completed your first pipeline! High five yourself and breathe a sigh of relief. :)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R (SageMath)",
   "language": "r",
   "name": "ir-sage"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.2.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}